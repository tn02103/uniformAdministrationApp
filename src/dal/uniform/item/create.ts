import { genericSAValidator } from "@/actions/validations";
import SaveDataException from "@/errors/SaveDataException";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { uniformNumberSchema } from "@/zod/uniform";
import { Prisma } from "@prisma/client";
import { z } from "zod";

const propSchema = z.object({
    numberMap: z.array(z.object({
        sizeId: z.string().uuid().or(z.enum(['amount'])),
        numbers: z.array(uniformNumberSchema),
    })),
    data: z.object({
        uniformTypeId: z.string().uuid(),
        generationId: z.string().uuid().optional(),
        comment: z.string(),
        active: z.boolean(),
    }),
});
type PropType = z.infer<typeof propSchema>

/**
 * Creates multiple UniformItems of a single type
 * @requires AuthRole.inspector
 * @param numbers Numbers of the UniformItem for each size. 
 *  If the Uniformtype does not use sizes, the Array has only one element with sizeId="amount"
 * @param data Data of the uniformItems that are to be created. The size of the Uniform is included in the param numbers
 * @returns number of created Items
 */
export const create = (props: PropType): Promise<number> => genericSAValidator(
    AuthRole.inspector,
    props,
    propSchema,
    {
        uniformTypeId: props.data.uniformTypeId,
        uniformGenerationId: props.data.generationId,
        uniformSizeId: props.numberMap.filter(n => n.sizeId !== "amount").map(n => n.sizeId)
    }
).then(([, { numberMap, data }]) => prisma.$transaction(async (client) => {
    // VALiDATE number
    // with existing
    const allNumbers = numberMap.reduce((arr: number[], value) => ([...arr, ...value.numbers]), []);
    const existingUniforms = await client.uniform.findMany({
        where: {
            fk_uniformType: data.uniformTypeId,
            number: { in: allNumbers },
            recdelete: null,
        }
    });
    if (existingUniforms.length > 0) {
        throw new SaveDataException("Number already in use")
    }

    // for duplications
    const uniqueSet = new Set(allNumbers);
    if (uniqueSet.size !== allNumbers.length) {
        throw new SaveDataException('some number is entered multiple times');
    };

    const type = await prisma.uniformType.findUniqueOrThrow({
        where: { id: data.uniformTypeId },
        include: {
            defaultSizelist: {
                include: { uniformSizes: true }
            },
            uniformGenerationList: {
                where: {
                    recdelete: null,
                },
                include: {
                    sizelist: { include: { uniformSizes: true } }
                }
            }
        }
    });

    // VALIDATE generation
    let sizelist = type.defaultSizelist;
    if (type.usingGenerations) {
        if (!data.generationId) {
            throw new SaveDataException('A genertion is required for this type');
        }
        const gen = type.uniformGenerationList.find(g => g.id === data.generationId);
        if (gen) {
            sizelist = gen.sizelist;
        } else {
            throw new SaveDataException('Not a valid genertion from the uniformType');
        }
    } else {
        if (data.generationId) {
            throw new SaveDataException('Type does not suport generations');
        }
    }

    // VALIDATE size
    let allowedSizes = [];
    if (!type.usingSizes) {
        allowedSizes = ["amount"];
    } else {
        if (!sizelist) {
            throw new SaveDataException('Could not create Uniformitems. Failed to find sizelist for selected type and generation');
        }
        allowedSizes = sizelist.uniformSizes.map(s => s.id);
    }
    if (!numberMap.every(map => allowedSizes.includes(map.sizeId))) {
        throw new SaveDataException("Not allowed size used");
    }

    return prisma.uniform.createMany({
        data: numberMap.reduce((arr: Prisma.UniformCreateManyInput[], map) => [
            ...arr,
            ...map.numbers.map(number => ({
                number,
                fk_uniformType: data.uniformTypeId,
                fk_generation: data.generationId,
                fk_size: (map.sizeId !== "amount") ? map.sizeId : null,
                comment: data.comment,
                active: data.active,
            })),
        ], []),
    }).then(d => d.count);
}));
