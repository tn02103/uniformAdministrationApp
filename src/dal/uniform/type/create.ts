import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { uniformTypeArgs } from "@/types/globalUniformTypes";
import { uniformTypeFormSchema, UniformTypeFormType } from "@/zod/uniformConfig";

export const create = (props: UniformTypeFormType) => genericSAValidator(
    AuthRole.materialManager,
    props,
    uniformTypeFormSchema,
    { uniformSizelistId: props.fk_defaultSizelist }
).then(async ([{ assosiation }, data]) => prisma.$transaction(async (client) => {
    const nameDuplication = await client.uniformType.findFirst({
        where: {
            fk_assosiation: assosiation,
            recdelete: null,
            name: props.name,
        }
    });
    if (nameDuplication) {
        return {
            error: {
                message: "custom.uniform.type.nameDuplication",
                formElement: "name",
            }
        };
    }

    const acronymDuplication = await client.uniformType.findFirst({
        where: {
            fk_assosiation: assosiation,
            recdelete: null,
            acronym: props.acronym,
        }
    });
    if (acronymDuplication) {
        return {
            error: {
                message: "custom.uniform.type.acronymDuplication;name:" + acronymDuplication.name,
                formElement: "acronym",
            }
        };
    }

    if (props.usingSizes && !props.fk_defaultSizelist) {
        return {
            error: {
                message: "pleaseSelect",
                formElement: "fk_defaultSizelist"
            }
        };
    }

    const amount = await client.uniformType.count({
        where: {
            fk_assosiation: assosiation,
            recdelete: null,
        }
    });

    return client.uniformType.create({
        ...uniformTypeArgs,
        data: {
            ...data,
            fk_assosiation: assosiation,
            sortOrder: amount,
        }
    });
}));
