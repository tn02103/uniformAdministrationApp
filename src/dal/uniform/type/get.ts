"use server"
import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { uniformTypeArgs } from "@/types/globalUniformTypes";
import { z } from "zod";

export const getUniformType = (props: string) => genericSAValidator(
    AuthRole.user,
    props,
    z.string().uuid(),
).then(([typeId, { assosiation }]) => prisma.uniformType.findUnique({
    where: {
        id: typeId,
        fk_assosiation: assosiation,
    },
    ...uniformTypeArgs
}));


