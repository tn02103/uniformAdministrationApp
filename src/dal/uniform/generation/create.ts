import { genericSAValidator } from "@/actions/validations";
import { SAReturnType } from "@/dal/_helper/testHelper";
import SaveDataException from "@/errors/SaveDataException";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { UniformType } from "@/types/globalUniformTypes";
import { uniformGenerationFormSchema } from "@/zod/uniformConfig";
import { z } from "zod";
import { __unsecuredGetUniformTypeList } from "../type/get";

const propSchema = uniformGenerationFormSchema.extend({
    uniformTypeId: z.string().uuid(),
});
type PropType = z.infer<typeof propSchema>;

export const create = (props: PropType): SAReturnType<UniformType[]> => genericSAValidator(
    AuthRole.materialManager,
    props,
    propSchema,
    { uniformTypeId: props.uniformTypeId, uniformSizelistId: props.fk_sizelist }
).then(([{ assosiation }, data]) => prisma.$transaction(async (client) => {
    const type = await client.uniformType.findUniqueOrThrow({
        where: {
            id: data.uniformTypeId,
        }
    });
    if (!type.usingGenerations) {
        throw new SaveDataException('generations are not activated for uniformType');
    }
    if (type.usingSizes && !props.fk_sizelist) {
        throw new SaveDataException('fk_sizelist is required for this uniformType');
    }
    else if (!type.usingSizes) {
        data.fk_sizelist = null;
    }

    const generationList = await client.uniformGeneration.findMany({
        where: {
            fk_uniformType: data.uniformTypeId,
            recdelete: null
        }
    });
    if (generationList.find(g => g.name === data.name)) {
        return {
            error: {
                message: "custom.uniform.generation.nameDuplication",
                formElement: "name",
            }
        }
    }

    await client.uniformGeneration.create({
        data: {
            name: data.name,
            outdated: data.outdated,
            fk_sizelist: data.fk_sizelist,
            fk_uniformType: data.uniformTypeId,
            sortOrder: generationList.length,
        }
    });
    return __unsecuredGetUniformTypeList(assosiation, client);
}));
