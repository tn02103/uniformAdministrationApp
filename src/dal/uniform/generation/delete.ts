import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { z } from "zod";
import { __unsecuredGetUniformTypeList } from "../type/get";

export const markDeleted = (props: string) => genericSAValidator(
    AuthRole.materialManager,
    props,
    z.string().uuid(),
    { uniformGenerationId: props }
).then(([{ assosiation, username }, id]) => prisma.$transaction(async (client) => {
    const gen = await client.uniformGeneration.findUniqueOrThrow({
        where: { id }
    });

    // TODO is this realy the correct way to handle this. 
    // I think the user should maby even decide if all the uniformItems from the Generations schould be deleted 
    // or the generation just set to null. The generation of a uniformItem normaly doesnt change 
    // and its more likely that the itmes should be deleted to.
    await client.uniform.updateMany({
        where: {
            recdelete: null,
            fk_generation: id,
        },
        data: {
            fk_generation: null,
        }
    });
    await client.uniformGeneration.update({
        where: { id },
        data: {
            recdelete: new Date(),
            recdeleteUser: username,
        }
    });
    await client.uniformGeneration.updateMany({
        where: {
            fk_uniformType: gen.fk_uniformType,
            recdelete: null,
            sortOrder: { gt: gen.sortOrder }
        },
        data: {
            sortOrder: { decrement: 1 }
        }
    });

    return __unsecuredGetUniformTypeList(assosiation, client);
}));
