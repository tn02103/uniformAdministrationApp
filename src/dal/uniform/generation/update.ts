"use server";

import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { UniformType } from "@/types/globalUniformTypes";
import { uniformGenerationFormSchema } from "@/zod/uniformConfig";
import { z } from "zod";
import { __unsecuredGetUniformTypeList } from "../type/get";


type ReturnType = Promise<{
    error: {
        message: string,
        formElement: string,
    }
} | UniformType[]>
const propSchema = z.object({
    data: uniformGenerationFormSchema,
    id: z.string().uuid(),
});
type PropType = z.infer<typeof propSchema>;

export const updateUniformGeneration = (props: PropType): ReturnType => genericSAValidator(
    AuthRole.materialManager,
    props,
    propSchema,
    { uniformGenerationId: props.id, uniformSizelistId: props.data.fk_sizelist }
).then(([{ assosiation }, { id, data }]) => prisma.$transaction(async (client) => {
    const type = await client.uniformType.findFirstOrThrow({
        where: {
            fk_assosiation: assosiation,
            uniformGenerationList: {
                some: { id }
            }
        }
    });

    const list = await client.uniformGeneration.findMany({
        where: {
            fk_uniformType: type.id,
            recdelete: null,
        }
    });
    if (list.find(g => g.id !== id && g.name === data.name)) {
        return {
            error: {
                message: "custom.uniform.generation.nameDuplication",
                formElement: "name",
            }
        }
    }
    
    if (!type.usingSizes) {
        data.fk_sizelist = null;
    } else if (!data.fk_sizelist) {
        return {
            error: {
                message: "pleaseSelect",
                formElement: "fk_sizelist"
            }
        }
    }

    await client.uniformGeneration.update({
        where: {id},
        data
    });

    return __unsecuredGetUniformTypeList(assosiation, client);
}));
