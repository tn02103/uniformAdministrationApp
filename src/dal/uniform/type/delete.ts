"use server";

import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { UniformType } from "@/types/globalUniformTypes";
import { Prisma, PrismaClient } from "@prisma/client";
import { z } from "zod";
import { __unsecuredGetUniformTypeList } from "./get";

export const markDeleted = (props: string): Promise<UniformType[]> => genericSAValidator(
    AuthRole.materialManager,
    props,
    z.string().uuid(),
    { uniformTypeId: props }
).then(([{ assosiation, username }, typeId]): Promise<UniformType[]> => prisma.$transaction(async (client) => {
    await Promise.all([
        returnAllUniformItems(typeId, client),
        deleteAllUniformItems(typeId, username, client),
        deleteAllGenerations(typeId, username, client),
    ]);

    const type = await deleteType(typeId, username, client as PrismaClient);
    await moveSortOrderUp(assosiation, type.sortOrder, client as PrismaClient);

    return __unsecuredGetUniformTypeList(assosiation, client);
}));

const returnAllUniformItems = (fk_uniformType: string, client: Prisma.TransactionClient) =>
    client.uniformIssued.updateMany({
        where: {
            uniform: { fk_uniformType },
            dateReturned: null,
        },
        data: {
            dateReturned: new Date(),
        },
    });

const deleteAllUniformItems = (fk_uniformType: string, username: string, client: Prisma.TransactionClient) =>
    client.uniform.updateMany({
        where: {
            fk_uniformType,
            recdelete: null,
        },
        data: {
            recdelete: new Date(),
            recdeleteUser: username,
        }
    });
const deleteAllGenerations = (fk_uniformType: string, username: string, client: Prisma.TransactionClient) =>
    client.uniformGeneration.updateMany({
        where: {
            fk_uniformType,
            recdelete: null
        },
        data: {
            recdelete: new Date(),
            recdeleteUser: username,
        },
    });

const deleteType = (id: string, username: string, client: Prisma.TransactionClient) =>
    client.uniformType.update({
        where: { id },
        data: {
            recdelete: new Date(),
            recdeleteUser: username,
        }
    });

const moveSortOrderUp = (fk_assosiation: string, minSortOrder: number, client: Prisma.TransactionClient) =>
    client.uniformType.updateMany({
        where: {
            fk_assosiation,
            sortOrder: { gt: minSortOrder },
            recdelete: null,
        },
        data: {
            sortOrder: { decrement: 1 }
        }
    });
