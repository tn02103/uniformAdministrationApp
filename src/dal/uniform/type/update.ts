"use server";

import { genericSAValidator } from "@/actions/validations";
import { AuthRole } from "@/lib/AuthRoles";
import { prisma } from "@/lib/db";
import { UniformType } from "@/types/globalUniformTypes";
import { uniformTypeFormSchema, UniformTypeFormType } from "@/zod/uniformConfig";
import { __unsecuredGetUniformTypeList } from "./get";
import { SAReturnType } from "@/dal/_helper/testHelper";

export const update = (props: UniformTypeFormType): SAReturnType<UniformType[]> => genericSAValidator(
    AuthRole.materialManager,
    props,
    uniformTypeFormSchema, // uniformTypeValidator.test(data),
    { uniformTypeId: props.id, uniformSizelistId: props.fk_defaultSizelist }
).then(([{ assosiation }, data]) => prisma.$transaction(async (client) => {
    const list = await client.uniformType.findMany({
        where: {
            fk_assosiation: assosiation,
            recdelete: null,
        }
    });

    if (list.find(t => t.id !== data.id && t.name === data.name)) {
        return {
            error: {
                message: "custom.uniform.type.nameDuplication",
                formElement: "name",
            }
        };
    }

    const acronymDupl = list.find(t => t.id !== data.id && t.acronym === data.acronym);
    if (acronymDupl) {
        return {
            error: {
                message: "custom.uniform.type.acronymDuplication;name:" + acronymDupl.name,
                formElement: "acronym",
            }
        };
    }
    if (data.usingSizes && !data.fk_defaultSizelist) {
        return {
            error: {
                message: "pleaseSelect",
                formElement: "fk_defaultSizelist"
            }
        };
    }

    await client.uniformType.update({
        where: {
            id: data.id,
        },
        data: data
    });

    return __unsecuredGetUniformTypeList(assosiation, client);
}));
