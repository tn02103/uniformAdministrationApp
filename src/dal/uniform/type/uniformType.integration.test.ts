/* eslint-disable jest/no-conditional-expect */
import { cleanDataV2, runServerActionTest } from "@/dal/_helper/testHelper";
import { prisma } from "@/lib/db";
import { UniformType } from "@/types/globalUniformTypes";
import { checkDateTolerance } from "../../../../jest/helpers/test-utils";
import { StaticData } from "../../../../tests/_playwrightConfig/testData/staticDataLoader";
import { create } from "./create";
import { markDeleted } from "./delete";
import { getList, getType } from "./get";
import { update } from "./update";

const { ids, cleanup, fk_assosiation, data } = new StaticData(0);

describe('<UniformType> Integration Tests', () => {
    afterEach(async () => {
        await cleanup.uniformTypeConfiguration();
    });

    it('should create, read, update uniform type end-to-end', async () => {
        // Test create
        const createProps = {
            name: 'Type Test',
            acronym: 'IT',
            issuedDefault: 2,
            usingSizes: false,
            usingGenerations: true,
            fk_defaultSizelist: null,
        };

        const { result: createResult, success: createSuccess } = await runServerActionTest(create(createProps));
        expect(createSuccess).toBeTruthy();
        expect(createResult).toBeDefined();
        expect(createResult).toEqual(
            expect.objectContaining(createProps)
        );

        // Test read single
        const { result: getResult, success: getSuccess } = await runServerActionTest(getType(createResult.id));
        expect(getSuccess).toBeTruthy();
        expect(getResult).toMatchObject({
            name: 'Type Test',
            acronym: 'IT',
            issuedDefault: 2,
        });

        // Test update
        const updateProps = {
            id: createResult.id,
            data: {
                ...createProps,
                name: 'Updated',
                issuedDefault: 3,
            }
        };

        const { result: updateResult, success: updateSuccess } = await runServerActionTest(update(updateProps));
        expect(updateSuccess).toBeTruthy();
        expect(updateResult).toBeDefined();

        const updatedType = updateResult?.find((t: { id: string }) => t.id === createResult.id);
        expect(updatedType).toMatchObject({
            name: 'Updated',
            issuedDefault: 3,
        });
    });

    it('should get list with correct data structure', async () => {
        const { result, success } = await runServerActionTest(getList());
        expect(success).toBeTruthy();

        expect(cleanDataV2(result)).toMatchSnapshot();
    });

    it('should get single type with correct data structure', async () => {
        const { result, success } = await runServerActionTest(getType(ids.uniformTypeIds[0]));
        expect(success).toBeTruthy();
        expect(cleanDataV2(result)).toMatchSnapshot();
    });

    it('should handle delete with cascading operations', async () => {
        // This tests the complex delete operation that involves multiple database operations
        const { success, result } = await runServerActionTest(markDeleted(ids.uniformTypeIds[1]));
        expect(success).toBeTruthy();
        expect((result as UniformType[]).filter((t) => t.id === ids.uniformTypeIds[1])).toHaveLength(0);

        const [type, generations, otherGenerations, uniformItems, otherUniformItems] = await prisma.$transaction([
            prisma.uniformType.findFirst({
                where: {
                    id: ids.uniformTypeIds[1],
                }
            }),
            prisma.uniformGeneration.findMany({
                where: {
                    fk_uniformType: ids.uniformTypeIds[1],
                }
            }),
            prisma.uniformGeneration.findMany({
                where: {
                    fk_uniformType: { not: ids.uniformTypeIds[1] },
                    type: { fk_assosiation },
                }
            }),
            prisma.uniform.findMany({
                where: {
                    fk_uniformType: ids.uniformTypeIds[1],
                }
            }),
            prisma.uniform.findMany({
                where: {
                    fk_uniformType: { not: ids.uniformTypeIds[1] },
                    type: { fk_assosiation },
                }
            }),
        ]);

        expect(type?.recdelete).not.toBeNull();
        expect(checkDateTolerance(type!.recdelete!)).toBeLessThan(5000);
        expect(type?.recdeleteUser).toEqual("mana");
        generations.forEach((gen) => {
            if (gen.id === ids.uniformGenerationIds[6]) {
                expect(gen.recdelete).not.toBeNull();
                expect(checkDateTolerance(gen.recdelete!)).toBeGreaterThan(5000);
                expect(gen.recdeleteUser).toEqual("test4");
            } else {
                expect(gen.recdelete).not.toBeNull();
                expect(checkDateTolerance(gen.recdelete!)).toBeLessThan(5000);
                expect(gen.recdeleteUser).toEqual("mana");
            }
        });
        otherGenerations.forEach((gen) => {
            expect(gen.recdelete === null || checkDateTolerance(gen.recdelete) > 5000).toBeTruthy();
        });
        const prevDeletedItems = uniformItems.filter((item) => data.uniformList.find((uni) => (uni.id === item.id && uni.recdelete)));
        const newDeletedItems = uniformItems.filter((item) => data.uniformList.find((uni) => (uni.id === item.id && !uni.recdelete)));
        prevDeletedItems.forEach((item) => {
            expect(item.recdelete).not.toBeNull();
            expect(checkDateTolerance(item.recdelete!)).toBeGreaterThan(5000);
        });
        newDeletedItems.forEach((item) => {
            expect(item.recdelete).not.toBeNull();
            expect(checkDateTolerance(item.recdelete!)).toBeLessThan(5000);
            expect(item.recdeleteUser).toEqual("mana");
        });
        otherUniformItems.forEach((item) => {
            expect(item.recdelete === null || checkDateTolerance(item.recdelete) > 5000).toBeTruthy();
        });
    });
});
