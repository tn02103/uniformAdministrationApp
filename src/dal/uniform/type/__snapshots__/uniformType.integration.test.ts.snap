// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UniformType> Integration Tests should get list with correct data structure 1`] = `
[
  {
    "acronym": "AA",
    "defaultSizelist": {
      "id": "242541-baa0-4303-be44-6c54898ca869",
      "name": "Liste1",
    },
    "fk_defaultSizelist": "242541-baa0-4303-be44-6c54898ca869",
    "id": "659ad7-7b45-4c31-8ee0-8f8c2d385886",
    "issuedDefault": 3,
    "name": "Typ1",
    "sortOrder": 0,
    "uniformGenerationList": [
      {
        "fk_sizelist": "242541-baa0-4303-be44-6c54898ca869",
        "id": "5a8f39-6e54-46f7-978b-cf23c52c042c",
        "name": "Generation1-1",
        "outdated": true,
        "sizelist": {
          "id": "242541-baa0-4303-be44-6c54898ca869",
          "name": "Liste1",
        },
        "sortOrder": 0,
      },
      {
        "fk_sizelist": "242541-baa0-4303-be44-6c54898ca869",
        "id": "1e05b1-740f-4f66-beea-aa39b37a3fe8",
        "name": "Generation1-2",
        "outdated": false,
        "sizelist": {
          "id": "242541-baa0-4303-be44-6c54898ca869",
          "name": "Liste1",
        },
        "sortOrder": 1,
      },
      {
        "fk_sizelist": "598ef5-1820-4a08-992c-2c3b91e8f093",
        "id": "4095a7-ff83-4a0f-91e5-75d3dfa28079",
        "name": "Generation1-3",
        "outdated": false,
        "sizelist": {
          "id": "598ef5-1820-4a08-992c-2c3b91e8f093",
          "name": "Liste2",
        },
        "sortOrder": 2,
      },
      {
        "fk_sizelist": "41b4a1-2f60-468e-9eba-8355cb7970fb",
        "id": "2f0541-7c4c-46c1-a03f-72ac6f4246d7",
        "name": "Generation1-4",
        "outdated": false,
        "sizelist": {
          "id": "41b4a1-2f60-468e-9eba-8355cb7970fb",
          "name": "Liste3",
        },
        "sortOrder": 3,
      },
    ],
    "usingGenerations": true,
    "usingSizes": true,
  },
  {
    "acronym": "AB",
    "defaultSizelist": null,
    "fk_defaultSizelist": null,
    "id": "eb645e-1bb4-4046-825a-16388265a803",
    "issuedDefault": 1,
    "name": "Typ2",
    "sortOrder": 1,
    "uniformGenerationList": [
      {
        "fk_sizelist": null,
        "id": "782ba5-f4ff-4829-97e8-b8923f870827",
        "name": "Generation2-1",
        "outdated": true,
        "sizelist": null,
        "sortOrder": 0,
      },
      {
        "fk_sizelist": null,
        "id": "892f4b-631e-4e01-aee8-0b1030810114",
        "name": "Generation2-2",
        "outdated": false,
        "sizelist": null,
        "sortOrder": 1,
      },
    ],
    "usingGenerations": true,
    "usingSizes": false,
  },
  {
    "acronym": "AC",
    "defaultSizelist": {
      "id": "598ef5-1820-4a08-992c-2c3b91e8f093",
      "name": "Liste2",
    },
    "fk_defaultSizelist": "598ef5-1820-4a08-992c-2c3b91e8f093",
    "id": "136e0b-3a30-4552-9296-dd1d51b814d2",
    "issuedDefault": 1,
    "name": "Typ3",
    "sortOrder": 2,
    "uniformGenerationList": [],
    "usingGenerations": false,
    "usingSizes": true,
  },
  {
    "acronym": "AD",
    "defaultSizelist": null,
    "fk_defaultSizelist": null,
    "id": "f93bf3-4aea-4d1a-b38f-a1a2f6b8970f",
    "issuedDefault": 1,
    "name": "Typ4",
    "sortOrder": 3,
    "uniformGenerationList": [],
    "usingGenerations": false,
    "usingSizes": false,
  },
]
`;

exports[`<UniformType> Integration Tests should get single type with correct data structure 1`] = `
{
  "acronym": "AA",
  "defaultSizelist": {
    "id": "242541-baa0-4303-be44-6c54898ca869",
    "name": "Liste1",
  },
  "fk_defaultSizelist": "242541-baa0-4303-be44-6c54898ca869",
  "id": "659ad7-7b45-4c31-8ee0-8f8c2d385886",
  "issuedDefault": 3,
  "name": "Typ1",
  "sortOrder": 0,
  "uniformGenerationList": [
    {
      "fk_sizelist": "242541-baa0-4303-be44-6c54898ca869",
      "id": "5a8f39-6e54-46f7-978b-cf23c52c042c",
      "name": "Generation1-1",
      "outdated": true,
      "sizelist": {
        "id": "242541-baa0-4303-be44-6c54898ca869",
        "name": "Liste1",
      },
      "sortOrder": 0,
    },
    {
      "fk_sizelist": "242541-baa0-4303-be44-6c54898ca869",
      "id": "1e05b1-740f-4f66-beea-aa39b37a3fe8",
      "name": "Generation1-2",
      "outdated": false,
      "sizelist": {
        "id": "242541-baa0-4303-be44-6c54898ca869",
        "name": "Liste1",
      },
      "sortOrder": 1,
    },
    {
      "fk_sizelist": "598ef5-1820-4a08-992c-2c3b91e8f093",
      "id": "4095a7-ff83-4a0f-91e5-75d3dfa28079",
      "name": "Generation1-3",
      "outdated": false,
      "sizelist": {
        "id": "598ef5-1820-4a08-992c-2c3b91e8f093",
        "name": "Liste2",
      },
      "sortOrder": 2,
    },
    {
      "fk_sizelist": "41b4a1-2f60-468e-9eba-8355cb7970fb",
      "id": "2f0541-7c4c-46c1-a03f-72ac6f4246d7",
      "name": "Generation1-4",
      "outdated": false,
      "sizelist": {
        "id": "41b4a1-2f60-468e-9eba-8355cb7970fb",
        "name": "Liste3",
      },
      "sortOrder": 3,
    },
  ],
  "usingGenerations": true,
  "usingSizes": true,
}
`;
